perf(ast/estree): `ESTree` serializer use `CodeBuffer`

overlookmotel opened the pull request:
`ESTree` serializer use `CodeBuffer` to assemble JSON string. `CodeBuffer` has a more efficient `push` implementation than `Vec<u8>`.

overlookmotel commented:
* **#9331** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9331?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9331?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9331).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Boshen commented:
No difference in performance?

overlookmotel commented:
Only +0.5% improvement. Damn, I hoped for more.

I guess the improvement in perf of `push` is insignificant in context of the rest of the work the serializer does.

Boshen commented:
Shall we reject the PR given the added dependency? `oxc_data_structures` will grow in size ... hence slower compile time overall due to `oxc_ast` -> `oxc_estree` -> `oxc_data_structures`.

overlookmotel commented:
Personally I still think it's worthwhile, for 3 reasons:

1. `CodeBuffer` can be optimized further, and it'd be ideal to do that in one place, so both `ESTree` serialization and `oxc_codegen` both benefit.
2. This PR removes near-identical code from `oxc_estree` (`Buffer` was same thing as `CodeBuffer`, just with less thorough doc comments!).
3. There's a sizeable perf improvement in this PR to generation of pretty-printed JSON. We won't see that in benchmarks as they only use compact JSON serializer, but it should gain us a little bit on time to run conformance.

However, you're absolutely right about the dependency chain. I had overlooked that `oxc_data_structures` can be an optional dependency for `oxc_estree`. Have changed that now. So, unless `serialize` feature is enabled, it adds no dependencies to `oxc_ast`. (#9338 also makes `itoa` dependency optional - I'd also missed that before)

But I've moved this PR to the top of the stack, so that we can drop it if you feel strongly.

overlookmotel commented:
(see also next PR #9340 which has a much bigger perf impact)

overlookmotel committed:
perf(ast/estree): `ESTree` serializer use `CodeBuffer` (#9331)

`ESTree` serializer use `CodeBuffer` to assemble JSON string. `CodeBuffer` has a more efficient `push` implementation than `Vec<u8>`.
